File: leet20/leet20.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashmap = {")": "(", "}": "{", "]": "["}
        stack = []

        for i in s:
            if i not in hashmap:
                stack.append(i)
            else:
                if not stack or hashmap[i] != stack.pop():
                    return False
        return not stack

File: leet20/leet20.cpp
#include <string>
#include <unordered_map>
#include <stack>

class Solution {
public:
    bool isValid(std::string s) {
        std::unordered_map<char, char> hashmap = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        std::stack<char> stack;

        for (char i : s) {
            if (hashmap.find(i) == hashmap.end()) {
                stack.push(i);
            } else {
                if (stack.empty() || stack.top() != hashmap[i]) {
                    return false;
                }
                stack.pop();
            }
        }

        return stack.empty();
    }
};
